### PR TITLE
[SYCLomatic] fix pointer to device memory bug.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -1193,6 +1193,7 @@ bool IterationSpaceBuiltinRule::renameBuiltinName(const DeclRefExpr *DRE,
   auto BuiltinName = DRE->getDecl()->getName();
   if (BuiltinName == "threadIdx")
     NewName = DpctGlobalInfo::getItem(DRE) + ".get_local_id(";
+  // here!!!
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: need delete the comment for debug.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 2]
```diff
@@ -11639,9 +11640,34 @@ void MemoryMigrationRule::registerMatcher(MatchFinder &MF) {
...
           .bind("unresolvedCall"),
       this);
+
+  MF.addMatcher(declRefExpr(to(varDecl(hasAttr(attr::CUDADevice), hasType(arrayType()) /*TODO: */)))
+                    .bind("declRefExpr"),
+                this);
 }
 
 void MemoryMigrationRule::runRule(const MatchFinder::MatchResult &Result) {
+  if (const auto *const DRE =
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: need extract a function.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 3]
```diff
@@ -11639,9 +11640,34 @@ void MemoryMigrationRule::registerMatcher(MatchFinder &MF) {
                            hasParent(callExpr(parentStmt()).bind("callExpr")))
           .bind("unresolvedCall"),
       this);
+
+  MF.addMatcher(declRefExpr(to(varDecl(hasAttr(attr::CUDADevice), hasType(arrayType()) /*TODO: */)))
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: TODO need be deleted.

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 4]
```diff
@@ -10262,6 +10262,22 @@ void MemVarRule::runRule(const MatchFinder::MatchResult &Result) {
...
+      if (const auto *arrType = MemVarRef->getType()->getAsArrayTypeUnsafe()) {
+        if (ICE->getCastKind() == CK_ArrayToPointerDecay &&
+            arrType->getElementType()->isArrayType() &&
+            MemVarRef->getDecl()->hasAttr<CUDADeviceAttr>() &&
+            isAssignOperator(getParentStmt(Parent))) {
+          std::string Replacement = "(" + ICE->getType().getAsString() + ")" +
+                                    MemVarRef->getDecl()->getName().str() +
+                                    ".get_ptr()";
+          emplaceTransformation(
+              new ReplaceStmt(MemVarRef, std::move(Replacement)));
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/ASTTraversal.cpp]: Use getDefinitionRange + ReplaceText

Comment (tangjj11) [reply clang/lib/DPCT/ASTTraversal.cpp]: done, change to these code:
```c++
std::string Replacement = "(" + ICE->getType().getAsString() + ")" +
                          MemVarRef->getDecl()->getName().str() +
                          ".get_ptr()";
auto Range = getDefinitionRange(MemVarRef->getBeginLoc(),
                                MemVarRef->getEndLoc());
auto &SM = DpctGlobalInfo::getSourceManager();
auto Length = Lexer::MeasureTokenLength(
    Range.getEnd(), SM,
    dpct::DpctGlobalInfo::getContext().getLangOpts());
auto Begin = Range.getBegin();
auto End = Range.getEnd();
Length += SM.getDecomposedLoc(End).second -
          SM.getDecomposedLoc(Begin).second;
emplaceTransformation(new ReplaceText(Range.getBegin(), Length,
                                      std::move(Replacement)));
```
and I have a question: we need use `getDefinitionRange` and then use `MeasureTokenLength` and `getDecomposedLoc` to get the length, can we provide an API to do these repeat tasks?

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by ziranzha — APPROVED** [review]
LGTM

**Pull request merged**